api/src/main/java/javax/servlet/http/HttpServlet.java

```diff
@@ -137,8 +137,15 @@ public HttpServlet() {
      * @see javax.servlet.ServletResponse#setContentType
      */
     protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        doMethod(req, resp);
+    }
+
+    // resolve method_not_supported message at runtime
+    private void doMethod(HttpServletRequest req, HttpServletResponse resp) throws IOException {
         String protocol = req.getProtocol();
-        String msg = lStrings.getString("http.method_get_not_supported");
+        String errMsg = lStrings.getString("http.method_not_supported");
+        String method = req.getMethod();
+        String msg = MessageFormat.format(errMsg, method);
         if (protocol.endsWith("1.1")) {
```

Comment:
Actually this logic is now not so good with HTTP/2.  It is probably only HTTP/1.0 that BAD_REQUEST should be sent and all others should be METHOD_NOT_ALLOWED.... Actaully I could argue that we could just always send METHOD_NOT_ALLOWED as while that was not defined in RFC1945, it was still a meaningful 4xx response code?  @stuartwdouglas your thoughts?

Reply:
I would just change it so that only HTTP/1.0 sends BAD_REQUEST. I don't think we should be modifying the existing 1.0 behaviour, its probably harmless but it's possible it could cause problems for someone.

Reply:
Note that there are probably some HTTP/0.9 interactions still happening... so we should also keep BAD_REQUEST for them too?

Reply:
I guess so. I don't feel strongly about this either way, but my thinking is keeping the existing behaviour is definitely fine, while changing it could cause problems (even though it is unlikely).

Reply:
How about:
```java
  switch(req.getProtocol())
  {
    case "HTTP/0.9":
    case "HTTP/1.1":
      resp.sendError(HttpServletResponse.SC_BAD_REQUEST, msg);
      break;
    default:
      resp.sendError(HttpServletResponse.SC_METHOD_NOT_ALLOWED, msg);
  }
```